rbac/providers/ldap/ldap_outbound_sync.py

```diff
@@ -0,0 +1,369 @@
+        throw ValueError.
+    """
+
+    LOGGER.info("Publishing to ldap: %s", outbound_entry)
+
+    payload_data_value = outbound_entry["data"]
+
+    server = Server(
+        LDAP_SERVER, get_info=ALL, connect_timeout=LDAP_CONNECT_TIMEOUT_SECONDS
+    )
+    ldap_connection = Connection(
+        server,
+        user=LDAP_USER,
+        password=LDAP_PASS,
+        receive_timeout=LDAP_RECEIVE_TIMEOUT_SECONDS,
+    )
+
+    try:
+        ldap_connection.bind()
+        connected_to_ldap = True
+    except LDAPSocketOpenError as lse:
+        raise LdapConnectionException(
+            "Failed to open a connection to Ldap. Aborting message transmission. Error: {0}".format(
+                lse
+            )
+        )
+
+    if connected_to_ldap:
+        # TODO: Should we remove the record if it fails? Move it to a retry queue?
+        LOGGER.info("Connected to ldap. Transmitting message, deleting record...")
+
+        try:
+            if outbound_entry["data_type"] == "user":
+                create_or_update_ldap_user(payload_data_value, ldap_connection)
+
+            elif outbound_entry["data_type"] == "group":
+                process_group_entry(payload_data_value, ldap_connection)
+
+            else:
+                # This case should be caught by the validator. But just in case...
+                LOGGER.warning(
+                    "Outbound queue record does not contain proper data type: %s",
+                    outbound_entry["data_type"],
+                )
+        except LDAPInvalidDnError as edn:
+            LOGGER.error("Encountered an error sending message to ldap. Error: %s", edn)
+
+
+def create_or_update_ldap_user(next_payload, ldap_connection):
+    """
+        Search AD to see if user from next exists. If an AD user exists, update the
+        AD user. If the AD user does not exist, create a new AD user.
+    """
+
+    distinguished_name = next_payload["distinguished_name"][0]
+
+    search_filter = USER_SEARCH_FILTER.format(distinguished_name)
+
+    LOGGER.info("Querying ldap for user...")
+    ldap_connection.search(LDAP_DC, search_filter, attributes=ldap3.ALL_ATTRIBUTES)
+
+    if ldap_connection.entries:
+        LOGGER.info("User found in ldap. Updating...")
+
+        ldap_payload = ldap_payload_transformer.to_user_create(next_payload)
+
+        for ad_attribute in ldap_payload:
+
+            operation = (MODIFY_REPLACE, [ldap_payload[ad_attribute][0]])
+            if ad_attribute == "member":
+                operation = (MODIFY_REPLACE, [ldap_payload["member"]])
+
+            ldap_connection.modify(
+                dn=distinguished_name, changes={ad_attribute: [operation]}
+            )
+
+        LOGGER.info("User updated in ldap")
+    else:
+        LOGGER.info("Creating user in ldap...")
+        _create_user_in_active_directory(next_payload, ldap_connection)
+        LOGGER.info("User created in ldap")
+
+
+def outbound_user_filter(next_user, provider):
```

Comment:
These functions are copied over from various parts of 'common' for incorporation into the architecture underway in 'ldap'. We discussed, and I have decided to leave 'common' alone to avoid breaking any functionality in 'azure'